.gitlab-ci.yml

```diff
@@ -25,12 +25,44 @@ stages:
   - deploy
   - notify
 
+########################################################################################################################
+# Configuration helpers
+########################################################################################################################
+
+.base-configuration:
+  tags: ['runner:main', 'size:large']
+  image: $CI_IMAGE
+
+.feature-branch-configuration:
+  extends: .base-configuration
+  except:
+    refs:
+      - main
+      - tags
+      - /^staging-[0-9]+$/
+      - /^release\/v[0-9.]+$/
+      - schedules
+
+.test-configuration:
+  stage: test
+  extends: .base-configuration
+  except:
+    refs:
+      - /^release\/v[0-9.]+$/
+      - schedules
```

Comment:
I found a bit hard to grasp currently since `test-configuration` and `feature-branch-configuration` are not comparable.
what about extracting configuration that only help us to select branches?
we could have configuration for feature branch, tag, main, ... and for each job specify on which kind of branch we want to run it.

something like:
```
format:
  extends: 
    - .main
    - .staging
    - .feature-branch
```
wdyt?

Reply:
I understand your point, I first tried this strategy but I ended up having the same amount of code while I wanted to shorten our configuration. 
Also you're example is sweet because the items in the extends section are of the same nature, but this example bellow is way less elegant and understandable:
```
format:
  extends: 
    - .base-configuration
    - .main
    - .staging
    - .potential-future-config
```

Reply:
fair enough, we could avoid composition among branches configuration and still extract branches/tag logic:

- `test-configuration` -> `test-allowed-branches` and remove `stage: test` and `.base-configuration`
- `feature-branch-configuration` -> `feature-branch`
- `tag` for  only refs tags
- `staging` for current staging

then we can have:
```
format:
  extends:
    - .base-configuration
    - .test-allowed-branches

notify-staging-failure:
  extends: 
    - .prepare_notification
    - .staging
```

Reply:
I'm not 100% convince of the benefit but why not. I will change `.staging` to `.staging-branch` though.